src/net/client/cache.rs

```diff
@@ -0,0 +1,1160 @@
+// if both AD and DO are not set in the original request then AD needs to be
+// cleared if it was set in the response (extra look up if no entry with
+// AD clear exists.
+//
+// The CD flag partitions the cache, responses to request with CD set must not
+// be visible to requests with CD and vice versa.
+//
+// A request with DO set can only be satisfied with a response to a request
+// with CD set. However, if CD in the request is clear then a reponse to a
+// request with CD set can be used if all unrequested DNSSEC records are
+// stripped.
+//
+// A request with RD clear can be satisfied by a response to a request with
+// RD set. For simplicitly requests with RD set can only get a response to
+// a request with RD set. In theory some responses to requests with RD clear
+// could satisfy requests with RD set.
+
+// Negative caching is described in RFC 2308.
+// NXDOMAIN and NODATA require special treatment. NXDOMAIN can be found
+// directly in the rcode field. NODATA is the condition where the answer
+// section does not contain any record that matches qtype and the message
+// is not a referral. NODATA is distinguished for a referral by the presence
+// of a SOA record in the authority section (a SOA record present implies
+// NODATA). A referral has one or more NS records in the authority section.
+// A NXDOMAIN reponse can only be cache if a SOA record is present in the
+// authority section. If the SOA record is absent then the NXDOMAIN response
+// should not be cached.
+// The TTL of the SOA record should reflect how long the response can be
+// cached. So no special treat is needed. Except that a different value should
+// limit the maximum time a negative response can be cached.
+//
+// Caching unreachable upstream should be limited to 5 minutes.
+// Caching SERVFAIL should be limited to 5 minutes.
+
+// RFC 8020 suggests a separate <QNAME, QCLASS> cache for NXDOMAIN, but
+// that may be too hard to implement.
+
+// RFC 9520 requires resolution failures to be cached for at least one
+// second. Resolution failure must not be cached for longer than 5 minutes.
+
+// RFC 8767 describes serving stale data.
+
+//------------ Config ---------------------------------------------------------
+
+/// Configuration of a cache.
+#[derive(Clone, Debug)]
+pub struct Config {
+    /// Maximum number of cache entries.
+    max_cache_entries: u64,
+
+    /// Maximum validity of a normal result.
+    max_validity: Duration,
+
+    /// Cache duration of transport failures.
+    transport_failure_duration: Duration,
+
+    /// Cache durations of misc. errors.
+    misc_error_duration: Duration,
+
+    /// Maximum validity of NXDOMAIN results.
+    max_nxdomain_validity: Duration,
+
+    /// Maximum validity of NODATA results.
+    max_nodata_validity: Duration,
+
+    /// Maximum validity of delegations.
+    max_delegation_validity: Duration,
+}
+
+impl Config {
+    /// Creates a new config with default values.
```

Comment:
What are the defaults? I think it would be good if they show up in the public RustDoc in some way.